soroban-env-host/src/native_contract/token/balance.rs

```diff
@@ -14,40 +12,40 @@ use soroban_env_common::{CheckedEnv, TryIntoVal};
 use super::error::ContractError;
 
 // Metering: *mostly* covered by components. Not sure about `try_into_val`.
-pub fn read_balance(e: &Host, id: Identifier) -> Result<BigInt, HostError> {
+pub fn read_balance(e: &Host, id: Identifier) -> Result<i128, HostError> {
     let key = DataKey::Balance(id);
     if let Ok(balance) = e.get_contract_data(key.try_into_val(e)?) {
-        Ok(balance.try_into_val(e)?)
+        balance.try_into_val(e)
     } else {
-        Ok(BigInt::from_u64(e, 0)?)
+        Ok(0)
     }
 }
 
 // Metering: *mostly* covered by components. Not sure about `try_into_val`.
-fn write_balance(e: &Host, id: Identifier, amount: BigInt) -> Result<(), HostError> {
+fn write_balance(e: &Host, id: Identifier, amount: i128) -> Result<(), HostError> {
     let key = DataKey::Balance(id);
     e.put_contract_data(key.try_into_val(e)?, amount.try_into_val(e)?)?;
     Ok(())
 }
 
 // Metering: covered by components.
-pub fn receive_balance(e: &Host, id: Identifier, amount: BigInt) -> Result<(), HostError> {
+pub fn receive_balance(e: &Host, id: Identifier, amount: i128) -> Result<(), HostError> {
     let balance = read_balance(e, id.clone())?;
     let is_frozen = read_state(e, id.clone())?;
     if is_frozen {
         Err(e.err_status_msg(ContractError::BalanceFrozenError, "balance is frozen"))
     } else {
-        write_balance(e, id, (balance + amount)?)
+        write_balance(e, id, balance + amount)
```

Comment:
Do we need overflow check (wrapping or checked_add) here?

Reply:
I remember you mentioned yesterday about turning on over/underflow protection that automatically converts them into a HostError?

Reply:
Good catch, yes, we do need to check this in the native (builtin) contract. Fixing.

(Actually I will only do _some_ of this because there's quite a lot of not-checked-correctly arithmetic in the builtin contract, this all needs to be done more carefully..)

On the guest (when running as WASM code) we do not, because guest code is going to be built with overflow checking turned on. It won't turn into a HostError directly; it'll turn into a WASM VM trap. But that will of course exit the VM invocation with a HostError.